商品購入機能

nana0817 opened the pull request:
# What
Furimaアプリに出品されている商品を購入できる機能。
# Why
Furimaアプリで商品を購入できるようにするため。

必要な情報を適切に入力して「購入」ボタンを押すと、商品の購入ができる動画
https://gyazo.com/43a520d7aa65756331d91760fc3b4144
入力に問題がある状態で「購入」ボタンが押された場合、情報は受け入れられず、購入ページでエラーメッセージが表示される動画
https://gyazo.com/0a90e544bc5892e90baeee2f5fe8c46f
ログイン状態の場合でも、URLを直接入力して自身が出品していない売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移する動画
https://gyazo.com/5ad0d52e52828c369872fa1011c56768
ログイン状態の場合でも、URLを直接入力して自身が出品した商品の商品購入ページに遷移しようとすると、商品の販売状況に関わらずトップページに遷移する動画
https://gyazo.com/099396bc96736296f28126ab5c8affde
ログアウト状態の場合は、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移する動画
https://gyazo.com/9bf4abba9419293e7a31f5f26f405811
トップページにて、売却済みの商品は、画像上に「sold out」の文字が表示される動画
https://gyazo.com/c4b4c51c96ce3face68ab45fd4877e37
商品詳細ページにて、売却済みの商品は、画像上に「sold out」の文字が表示される動画
https://gyazo.com/e25c897e49f62913df619d0678642332
ログイン状態の場合でも、売却済みの商品には、「商品の編集」「削除」「購入画面に進む」ボタンが表示されない動画
https://gyazo.com/28116ab85f795ffb60df7252ee319164
ログイン状態の場合でも、URLを直接入力して自身が出品した売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画
https://gyazo.com/fe864997254a9a4e6e9cf0926371cd69
テスト結果の画像
https://i.gyazo.com/a05ccbe800266bf5d0174a48b16d6738.png

nana0817 committed:
Formオブジェクトを作成

以下を実行
・app/modelsディレクトリにorder_address.rbファイルを作成
・モデルの一部機能を持たせるために「include ActiveModel::Model」を記述
・attr_accessorに保存したいカラム名を記述
・それぞれのカラムのバリデーションを記述
・saveメソッドの中にデータをテーブルに保存する処理を記述

nana0817 committed:
購入機能に必要なモデルやコントローラーを作成

以下を実行
・Orderモデルを作成
・Addressモデルを作成
・OrderモデルとAddressモデルのアソシエーションを定義
・OrderモデルとItemモデルのアソシエーションを定義
・ordersコントローラーを作成
・config/routes.rbにて、ordersのルーティングを記述(itemsのルーティングにネスト)
・ordersコントローラーにFormオブジェクトのインスタンスを生成する処理を記述

nana0817 committed:
ActiveHashを用いた各モデルの中にあるデータの修正

idが1からではなく、0から始まるようにした。

nana0817 committed:
購入のためのフォーム実装

以下を実行
・form_withにmodelオプションと、urlオプションを追加し、属性値を指定
・購入する商品の情報が表示されるよう編集
・各フォームのname属性の属性値を適切な値を変更
・エラーメッセージを表示する記述を追加

nana0817 committed:
商品詳細ページの「購入画面に進む」ボタンにorders#indexのパスを指定

nana0817 committed:
orders, addressesテーブルのマイグレーションファイルを編集し、マイグレート

nana0817 committed:
itemのFactoryBotの記述を変更

nana0817 committed:
OrderAdressモデルのテストコードの実装

以下を実行
１，テストファイルを作成
２，order_addressesのFactoryBot設定
３，テストがうまく実行されるようにconfig/environment/test.rbファイルを編集
４，ordersのFactoryBotにassociationメソッドを用いて、userとitemとのアソシエーションを定義
５，app/models/order_address.rbにて、
バリデーションを変更
６，テストコード実装

nana0817 committed:
クライアントサイドでPAY.JPのAPIを使用するために必要なJavaScriptを読み込めるようにする

furimaアプリでpay.jsを読み込めるようにした。

nana0817 committed:
決済機能のクライアントサイド実装(クレジットカード情報のトークン化)

以下を実装
１，トークン化を行うJavaScriptファイルを作成
２，作成したファイルが読み込めるようimportmap.rbとapplication.jsに記述を追加
３，card.jsにコードを記述

nana0817 committed:
Javascriptファイル読み込みの不具合を解消

nana0817 committed:
決済機能のクライアントサイド実装(サーバーサイドへのトークン送付)

nana0817 committed:
決済機能のサーバーサイド実装

以下を実行
・ストロングパラメーターを編集し、tokenの情報を取得できるようにした。
・order_address.rbにてattr_accessorの値にtokenを追加。order_idは削除
・Gem 'payjp'の導入
・orders_controller.rbに決済処理を記述
・order_address.rbにtokenのバリデーションを追加。order_idは削除
・FacotoryBotでtokenのダミーデータを作成できるよう変更
・テストコードにて「tokenが空では購入できない」というexampleを追加。order_idに関するものは削除

nana0817 committed:
サーバーサイドで環境変数を読み込めるように修正

nana0817 committed:
クライアントサイドで環境変数を読み込めるように

以下を実行
・Gem 'gon'を導入
・ビューでgonを使用できるよう、app/views/orders/index.html.erbに設定を追加
・orders_controller.rbにて、RailsからJavaScriptへ公開鍵を渡すためのコードを追加
・card.jsにて、JavaScriptで公開鍵を読み込むための記述を追加

nana0817 committed:
gonの導入に伴って、入力に不備があった場合の処理を追加

以下を実行
・orders_contoller.rbのcreateアクションのelseに、RailsからJavaScriptへ公開鍵を渡すコードを追加
・card.jsにて、renderが実行された場合でもJavaScriptのコードが実行されるよう編集

nana0817 committed:
購入機能の他関連機能を実装

以下を実行
・app/views/itemsディレクトリのindex.html.erb・show.thml.erbにて、売却済みの商品は、画像上に「sold out」の文字が表示されるように条件分岐を追加

・app/views/items/show.thml.erbにて、「編集」「削除」「購入」のボタンが、その商品が売却済みの場合に、表示しないよう条件分岐を追加

・items_controller.rbにて、出品者でも、すでに売却済みの商品の編集ページにアクセスできないようにプライベートメソッドを追加

nana0817 committed:
ActiveHashで管理するデータを取得する記述の変更

ActiveHashで管理するデータのidが０から始まるように変えたので、それに伴って、ビューファイルのデータを取得する記述も変更

nana0817 committed:
ordersコントローラーのindexアクション前に行うメソッドを追加

・ログインしていないユーザーが購入ページに遷移できないように、authenticate_user!メソッドを追加

・出品者が出品した商品の購入ページに遷移できないように、are_sellerメソッドを追加、その処理をプライベートメソッドとして記述

nana0817 committed:
Rubocopでコードを自動修正

koyamatakumi commented on app/models/order_address.rb:
```diff
@@ -0,0 +1,29 @@
+    validates :street_address
+    validates :token
+  end
+  validates :telephone_number, presence: true
```
```ruby
validates :telephone_number, presence: true
```
presence: trueを消して、with_options presence: true do~end内に記述を書きましょう。

koyamatakumi reviewed (commented):
nana0817様、レビュー依頼フォームの送信ありがとうございます。
コードを確認させていただき、コメントいたしましたのでご確認をお願いいたします。

修正後はコミット、プッシュを忘れずに行い、改めてレビュー依頼フォームの送信をお願い致します！

nana0817 commented:
コードレビューありがとうございます。
早速修正に取り掛かります！

nana0817 committed:
バリデーションの記述をリファクタリング

validates :telephone_number, presence: trueの記述を削除して、
with_options presence: true do ~ endの中に含めた

nana0817 committed:
ActiveHashのデータを取得する記述のリファクタリング

nana0817 committed:
Rubocopでコードを修正

nana0817 commented:
ご指摘いただいた点は全て修正しました！
再度コードレビューをお願いします！

OhyaYouki commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,45 @@
+
+  def are_seller
+    @item = Item.find(params[:item_id])
+    return unless current_user.id == @item.user_id
```
商品が購入済みである場合に購入ページに遷移できないような条件分岐がございません。
追加をお願いいたします。

OhyaYouki reviewed (commented):
レビュー依頼ありがとうございます。

コメントいたしましたので、ご確認をお願いいたします。

コードの修正後、コミット・プッシュを行った上で再度レビュー依頼をお送りください。
ご対応のほど、よろしくお願いいたします。

nana0817 committed:
ordersコントローラーに売却済みの商品の購入ページに遷移できないようにする条件分岐を追加、プライベートメソッドの名前を変更

nana0817 commented:
ご指摘いただいた点は全て修正しました！
再度コードレビューをお願いします！

maimai0401 commented on app/models/item.rb:
```diff
@@ -1,6 +1,7 @@
   # アソシエーション
   belongs_to :user
+  has_one :order
 
```
userモデルに購入管理テーブルとのアソシエーションが確認できませんでした。
追記しておきましょう。

maimai0401 reviewed (commented):
レビュー依頼ありがとうございます。プルリクエスト確認させていただきました。



修正点をコメントさせて頂きましたので、ご確認をお願い致します。

【次回プルリクエスト提出する際の注意点】
```ruby
①修正が終わりましたら、`コミットとプッシュ`をして、`プルリクエストに修正内容が反映されているか確認`しましょう。
②既に頂いている挙動確認の画像や動画が修正範囲だった場合、最新の画像や動画の提出をお願い致します。
③googleフォームの送信をお願いいたします。
```


ご不明点や疑問点ございましたら、チャット・コネクトメンターにご相談ください！
どうぞよろしくおねがいします。

nana0817 committed:
Orderモデルとのアソシエーションを記述

nana0817 commented:
ご指摘いただいた点を修正しました！
コードレビューをお願いします！

maimai0401 reviewed (commented):
レビュー依頼ありがとうございます。

こちら`LGTM`とさせていただきます！実装お疲れさまでした。

【注意】
```ruby
①（次のブランチを切る前に）プルリクエストのMergeとmainブランチへのPullを忘れないようにしましょう。
②次のフェーズに進む際は必ず新しいブランチを作ってから作業しましょう。
```

ご不明点や疑問点ございましたら、チャット・コネクトメンターにご相談ください！
引き続き頑張りましょう！応援しております！
